feat: support @httpApiKeyAuth trait

glb opened the pull request:
*Issue #, if available:* Fixes #453.

*Description of changes:* This adds codegen support for services and operations that use the `@httpApiKeyAuth` trait.

The generated client will have a new `apiKey` config attribute. This value will be injected into the header/query parameter that's specified in the trait by new middleware that's included in the codegen client.

⚠️ I've done my best to base this on existing patterns but I'm sure that there will be comments and things that need to change. Please let me know and I'll happily make adjustments.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

kstich commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @return true if the operation has the @optionalAuth trait
+     */
+    private static boolean operationUsesOptionalAuth(Model model, ServiceShape service, OperationShape operation) {
```
Doesn't feel like this needs a separate method either.

kstich commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @return true if the service has the @httpApiKeyAuth trait
+     */
+    private static boolean hasHttpApiKeyAuthTrait(ServiceShape service) {
```
It doesn't feel like this needs a separate method to check.

kstich commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java:
```diff
@@ -0,0 +1,178 @@
+                    // `in` and `name` are required attributes of the trait, `scheme` is optional
+                    put("in", s.getTrait(HttpApiKeyAuthTrait.class).get().getIn().toString());
+                    put("name", s.getTrait(HttpApiKeyAuthTrait.class).get().getName());
+                    put("scheme", s.getTrait(HttpApiKeyAuthTrait.class).get().getScheme().orElse(null));
```
This map is eventually [passed to `getFunctionParametersList`](https://github.com/awslabs/smithy-typescript/blob/379611e332bfcee9a4cb9d3c328089b647de0562/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/CodegenUtils.java#L147), which looks like it'll fail when receiving a `null`. There's not a test for a trait without a `scheme`, so that's a good thing to introduce either way. Since `getScheme` returns an `Optional`, this can use an `ifPresent` chain.

kstich commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java:
```diff
@@ -0,0 +1,178 @@
+                + "// https://github.com/awslabs/smithy-typescript/blob/main/smithy-typescript-codegen/"
+                + "src/main/resources/software/amazon/smithy/aws/typescript/codegen/integration/";
+
+        // write the middleware source
```
nit: comments should be complete sentences, starting with capitals and ending with periods.

This occurs in other places as well (including the test file.)

kstich commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java:
```diff
@@ -0,0 +1,178 @@
+    public List<RuntimeClientPlugin> getClientPlugins() {
+        return ListUtils.of(
+            // Add the config if the service uses HTTP API key authorization
+            RuntimeClientPlugin.builder()
```
formatting: conventionally, chained and continued lines are indented twice (8 extra spaces) and method invocations start on the next line.

```
RuntimeClientPlugin.builder()
        .inputConfig(Symbol.builder()
                .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
                .name("HttpApiKeyAuthInputConfig")
                .build())
        .resolvedConfig(...
```

This occurs in other places as well (including the test file.)

glb replied:
Thanks! I think I've fixed these but I may not have gotten them all right, please let me know if it's still not correctly formatted.

kstich commented on smithy-typescript-codegen/src/test/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPluginTest.java:
```diff
@@ -0,0 +1,150 @@
+            is(false));
+    }
+
+}
```
nit: files should end with an empty line.

Looks like this happens in a few of the test case files as well.

kstich commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java:
```diff
@@ -0,0 +1,178 @@
+ * Add config and middleware to support a service with the @httpApiKeyAuth trait.
+ */
+@SmithyInternalApi
+public final class AddHttpApiKeyAuthPlugin implements TypeScriptIntegration {
```
This doesn't look like it utilizes [the `@auth` trait](https://awslabs.github.io/smithy/1.0/spec/core/auth-traits.html#smithy-api-auth-trait) at all, but seems like it should. Note that [the `ServiceIndex`](https://awslabs.github.io/smithy/javadoc/1.15.0/software/amazon/smithy/model/knowledge/ServiceIndex.html) can help reason about the `@auth` trait in a clear way, notably [`getEffectiveAuthSchemes​(ToShapeId service, ToShapeId operation)`](https://awslabs.github.io/smithy/javadoc/1.15.0/software/amazon/smithy/model/knowledge/ServiceIndex.html#getEffectiveAuthSchemes(software.amazon.smithy.model.shapes.ToShapeId,software.amazon.smithy.model.shapes.ToShapeId)).

This will require additional tests for different effective schemes as well.

glb replied:
Thanks @kstich . I based this class on [AddAwsAuthPlugin.java](https://github.com/aws/aws-sdk-js-v3/blob/main/codegen/smithy-aws-typescript-codegen/src/main/java/software/amazon/smithy/aws/typescript/codegen/AddAwsAuthPlugin.java#L249) which also doesn't reference the `@auth` trait. I'll have to do some reading to understand how all this is meant to tie together, any further pointers you could provide would be greatly appreciated.

glb replied:
@kstich based on my reading what you're suggesting seems pretty straightforward, but confirmation would be awesome.

Based on your comment, I should be using `getEffectiveAuthSchemes(service)` or `getEffectiveAuthSchemes(service, operation)` and then checking the values of the returned map for the `HttpApiKeyAuthTrait`. Yes?

It doesn't look like `getEffectiveAuthSchemes` takes `@optionalAuth` into account, so I'm guessing I need to check for that myself, at least for now. I can write some tests to validate this but want to confirm the intended behaviour rather than what the code happens to implement today.

glb replied:
I've made an attempt at this, hopefully it's what you intended. The tests pass.

kstich replied:
Pretty close on this application, left a couple more comments 👍🏻

glb commented:
First round of fixes pushed, will rebase & squash them once I've got all the comments addressed.

glb commented:
@kstich @mtdowling @JordonPhillips what are the next steps I need to take for this PR? Is there any chance you'll have an opportunity to review soon? I don't mean to be pushy, just don't know what timeline to expect here as I haven't heard anything since Tuesday afternoon.

I left the changes from code review comments as separate commits so they would be easier to review, but I could rebase, squash back to a single commit, and mark ready for review. I think I've addressed all of Kevin's comments and the tests (unit tests as well as manual testing of a generated client) seem to indicate this does what I hoped it would.

glb commented:
Pushed a change (`f0ee2e5`) to delete a test; not 100% sure I understood the intended behaviour of the auth traits so backed off on this one as I discovered it was failing and it seemed like the expectation was incorrect. Would love to hear if the test was legitimately wrong or if there's a bug in the code.

kstich commented on smithy-typescript-codegen/src/test/resources/software/amazon/smithy/typescript/codegen/integration/http-api-key-auth-trait-all-optional.smithy:
```diff
@@ -0,0 +1,39 @@
+structure GetBarOutput {}
+
+@error("client")
+structure GetBarError {}
```
Can the `.smithy` files also get empty last lines?

glb commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java:
```diff
@@ -0,0 +1,159 @@
+        ServiceIndex index = ServiceIndex.of(model);
+
+        for (OperationShape operation : operations) {
+            if (index.getEffectiveAuthSchemes(service, operation).isEmpty()
```
In fact it cannot; it ends up at 127 characters.

glb commented on smithy-typescript-codegen/src/main/resources/software/amazon/smithy/typescript/codegen/integration/http-api-key-auth.ts:
```diff
@@ -0,0 +1,125 @@
+        ? `${options.scheme} ${resolvedConfig.apiKey}`
+        : resolvedConfig.apiKey;
+    } else if (options.in === "query") {
+      // It's terrifying that this is an option; API keys should not be part of the URL.
```
Fair; I was expecting both of these bits of feedback. I'll look through the other codegen to see how query parameters are supposed to be added to requests.

glb replied:
Done in eae855f.

glb commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java:
```diff
@@ -0,0 +1,159 @@
+
+        // If the service doesn't use HTTP API keys, we don't need to do anything and the generated
+        // code doesn't need any additional files.
+        if (!service.hasTrait(HttpApiKeyAuthTrait.class) || areAllOptionalAuthOperations(model, service)) {
```
👍 Extracting to a separate method as this is in 3 places.

glb replied:
I'm curious why `getEffectiveAuthSchemes` doesn't take `@optionalAuth` into account; it seems like it should?

glb commented:
Pushed changes to address most comments; still need to do the work to support an API key in a query parameter.

glb commented:
I think I've addressed all of the review comments so far. I've realized, though, that I'm missing the validator to check for conflicts. I probably won't get to that today, maybe tomorrow.

glb commented:
Silly me, I got confused, the validator is for a different PR. I think this is ready to be re-reviewed.

kstich commented:
I think this is ready to move to non-draft, I've initiated the CI run.

glb commented:
Thanks @kstich !

glb commented on smithy-typescript-codegen/src/main/resources/software/amazon/smithy/typescript/codegen/integration/http-api-key-auth.ts:
```diff
@@ -0,0 +1,122 @@
+  apiKey?: string;
+}
+
+export function resolveHttpApiKeyAuthConfig<T>(
+  input: T & PreviouslyResolved & HttpApiKeyAuthInputConfig
+): T & HttpApiKeyAuthResolvedConfig {
+  return input;
+}
```
OK, thanks... I assume that I also must delete https://github.com/awslabs/smithy-typescript/pull/473/files#diff-d568eb2f093fdfe09e49d3987c65f2a8022f43509666d8b4e3a00d8bd22ecdf6R67-R74 and will need to remove `HttpApiKeyAuthResolvedConfig` interface (etc etc) otherwise bad things will happen.

glb replied:
Tried removing `resolvedConfig` and `resolveFunction` as suggested and got the following error:

```
Projection source failed: java.lang.IllegalStateException: If any of inputConfig, resolvedConfig, or resolveFunction are set, then all of inputConfig, resolvedConfig, and resolveFunction must be set: inputConfig: SymbolReference{symbol=./src/middleware/HttpApiKeyAuth/HttpApiKeyAuthInputConfig, alias='HttpApiKeyAuthInputConfig', options=[DECLARE, USE]}, resolvedConfig: null, resolveFunction: null
software.amazon.smithy.typescript.codegen.integration.RuntimeClientPlugin.<init>(RuntimeClientPlugin.java:72)
...
```

Thoughts?

glb replied:
Restored `resolvedConfig` and `resolveFunction` (and the resolve function & related types) and things are working again.

trivikr replied:
You can just add pluginFunction and skip inputConfig/resolvedConfig/resolveFunction.

This falls under `HAS_MIDDLEWARE` Convention
https://github.com/awslabs/smithy-typescript/blob/d263078b81485a6a2013d243639c0c680343ff47/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/RuntimeClientPlugin.java#L753-L768

trivikr replied:
Posted suggestion for codegen changes in comment https://github.com/awslabs/smithy-typescript/pull/473#discussion_r774716937

glb replied:
I tried this and I get the following compile error when compiling the generated TypeScript:

```
src/AuditClient.ts:203:11 - error TS2345: Argument of type '{ runtime: string; base64Decoder: Decoder; base64Encoder: Encoder; bodyLengthChecker: (body: any) => number | undefined; defaultUserAgentProvider: Provider<UserAgent>; ... 17 more ...; customUserAgent?: string | ... 1 more ... | undefined; } & CustomEndpointsResolvedConfig & RetryResolvedConfig & HostHeaderResolvedC...' is not assignable to parameter of type 'AuditClientResolvedConfig'.
  Type '{ runtime: string; base64Decoder: Decoder; base64Encoder: Encoder; bodyLengthChecker: (body: any) => number | undefined; defaultUserAgentProvider: Provider<UserAgent>; ... 17 more ...; customUserAgent?: string | ... 1 more ... | undefined; } & CustomEndpointsResolvedConfig & RetryResolvedConfig & HostHeaderResolvedC...' is not assignable to type 'Required<ClientDefaults>'.
    Types of property 'apiKey' are incompatible.
      Type 'string | undefined' is not assignable to type 'string'.
        Type 'undefined' is not assignable to type 'string'.
```

I'm not sure, but this seems to be the problem:

```
type AuditClientResolvedConfigType = __SmithyResolvedConfiguration<__HttpHandlerOptions>
  & Required<ClientDefaults>
```

because it's making the `apiKey` field required instead of optional. I don't know what generates that code or why it's requiring all of the `ClientDefaults` fields to be present.

I think this is what took me down the path of a separate config type and not using `addConfigInterfaceFields` in the first place.

glb replied:
@trivikr happy new year, hope you had a chance to take some holidays and get refreshed! Any thoughts on a path forward here? I'm a bit stuck as I can't make the change you requested because the resulting generated code won't compile.

trivikr replied:
@glb Thank you for your patience. The PR is now merged.

glb commented:
Pushed mixup commits to address comments and fix missing `@jest/types` dependency so that the generated code can build & test.

trivikr commented on smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddHttpApiKeyAuthPlugin.java:
```diff
@@ -0,0 +1,169 @@
+    @Override
+    public List<RuntimeClientPlugin> getClientPlugins() {
+        return ListUtils.of(
+            // Add the config if the service uses HTTP API key authorization.
+            RuntimeClientPlugin.builder()
+                    .inputConfig(Symbol.builder()
+                            .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
+                            .name("HttpApiKeyAuthInputConfig")
+                            .build())
+                    .resolvedConfig(Symbol.builder()
+                            .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
+                            .name("HttpApiKeyAuthResolvedConfig")
+                            .build())
+                    .resolveFunction(Symbol.builder()
+                            .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
+                            .name("resolveHttpApiKeyAuthConfig")
+                            .build())
+                    .servicePredicate((m, s) -> hasEffectiveHttpApiKeyAuthTrait(m, s))
+                    .build(),
+
+            // Add the middleware to operations that use HTTP API key authorization.
+            RuntimeClientPlugin.builder()
+                    .pluginFunction(Symbol.builder()
+                            .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
```
Refs: https://github.com/awslabs/smithy-typescript/pull/473#discussion_r774711588
```suggestion
            // Add the middleware to operations that use HTTP API key authorization.
            RuntimeClientPlugin.builder()
                    .pluginFunction(Symbol.builder()
                            .namespace("./" + CodegenUtils.SOURCE_FOLDER + "/middleware/HttpApiKeyAuth", "/")
```

trivikr replied:
Add `apiKey` config by overriding `addConfigInterfaceFields` instead.

Example https://github.com/awslabs/smithy-typescript/blob/d263078b81485a6a2013d243639c0c680343ff47/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/AddChecksumRequiredDependency.java#L45-L69

glb replied:
Tried this, see resulting error message when trying to compile the generated client. https://github.com/awslabs/smithy-typescript/pull/473#discussion_r775072233

glb commented:
Pushed a behavioural change in `f2f40d3` so that the middleware doesn't throw an error when the API key is not available, because the middleware doesn't know if there is another auth mechanism available (for example if an operation can be authorized by an API key or a bearer token, it shouldn't fail because the API key isn't present).
